src/main/flight/pid.c

```diff
@@ -582,59 +585,61 @@ int8_t acroTrainerSign(float x)
     if (!FLIGHT_MODE(ANGLE_MODE) && !FLIGHT_MODE(HORIZON_MODE)) {
         bool resetIterm = false;
-        float projectedAngle;
-        const int8_t setpointSign = acroTrainerSign(setPoint);
+        float projectedAngle = 0;
+        const int setpointSign = acroTrainerSign(setPoint);
         const float currentAngle = (attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f;
-        const int8_t angleSign = acroTrainerSign(currentAngle);
+        const int angleSign = acroTrainerSign(currentAngle);
+
+        if ((acroTrainerAxisState[axis] != 0) && (acroTrainerAxisState[axis] != setpointSign)) {  // stick has reversed - stop limiting
+            acroTrainerAxisState[axis] = 0;
+        }
 
         // Limit and correct the angle when it exceeds the limit
-        if ((fabsf(currentAngle) > acro_trainer_angle_limit) || (acroTrainerAxisState[axis] != 0)) {
-            if (angleSign == -setpointSign) {  // stick has reversed - stop limiting
-                acroTrainerAxisState[axis] = 0;
-            } else {
-                if (acroTrainerAxisState[axis] != angleSign) {
-                    acroTrainerAxisState[axis] = angleSign;
-                    resetIterm = true;
-                }
-                ret = ((acro_trainer_angle_limit * angleSign) - currentAngle) * ACRO_TRAINER_GAIN;
+        if ((fabsf(currentAngle) > acroTrainerAngleLimit) && (acroTrainerAxisState[axis] == 0)) {
+            if (angleSign == setpointSign) {
+                acroTrainerAxisState[axis] = angleSign;
+                resetIterm = true;
             }
         }
+
+        if (acroTrainerAxisState[axis] != 0) {
+            ret = ((acroTrainerAngleLimit * angleSign) - currentAngle) * acroTrainerGain;
+        } else {
         
         // Not currently over the limit so project the angle based on current angle and
         // gyro angular rate using a sliding window based on gyro rate (faster rotation means larger window.
         // If the projected angle exceeds the limit then apply limiting to minimize overshoot.
-        if (acroTrainerAxisState[axis] == 0) {
-            const timeUs_t checkInterval = MAX(lrintf(constrainf(fabsf(gyro.gyroADCf[axis]) / 500.0f, 0.0f, 1.0f) * acro_trainer_lookahead_time), ACRO_TRAINER_LOOKAHEAD_MIN);
-            projectedAngle = (gyro.gyroADCf[axis] * checkInterval * 0.000001f) + currentAngle;
-            const int8_t projectedAngleSign = acroTrainerSign(projectedAngle);
-            if ((fabsf(projectedAngle) > acro_trainer_angle_limit) && (projectedAngleSign == setpointSign)) {
-                ret = ((acro_trainer_angle_limit * projectedAngleSign) - projectedAngle) * ACRO_TRAINER_GAIN;
+            // Calculate the lookahead window by scaling proportionally with gyro rate from 0-500dps
+            float checkInterval = constrainf(fabsf(gyro.gyroADCf[axis]) / ACRO_TRAINER_LOOKAHEAD_RATE_LIMIT, 0.0f, 1.0f) * acroTrainerLookaheadTime;
+            checkInterval = MAX(checkInterval, ACRO_TRAINER_LOOKAHEAD_MIN_TIME);
```

Comment:
(is MIN_TIME necessary ? Gyro rate must be small to trigger it, so using short lookahead should be OK - it will take some time to hit limit anyway)

Reply:
Yes, thanks for noticing this. This was the cause of the oscillation when approaching the angle limit very gently.

It was left over from a previous method I tried out that calculated the amount of time from now that an overflow was predicted (based on gyro rate and current angle).  It then used that in the denominator of a formula to calculate the new setpoint (if time is small then we need to push harder, etc.).  So I had to guard against division by zero and limit the effect for very small time intervals.  With the current logic the minimum is not needed and is detrimental.